Components: Make the `ProgressBar` public

fullofcaffeine opened the pull request:
Issue: https://github.com/WordPress/gutenberg/issues/59418.

## What?

This is part of a [larger effort](https://github.com/WordPress/gutenberg/issues/59418) to remove `__experimental` prefix from all "experimental" components, effectively promoting them to regular stable components. See the related issue for more context.

## Why?

The strategy of prefixing exports with `__experimental` has become deprecated after the introduction of private APIs. 

## How?

1. Export it from components without the `__experimental` prefix;
1. Keep the old `__experimental` export for backwards compatibility;
1. Change all imports of the old `__experimental` in GB and components to the one without the prefix (including in storybook stories). Also, update the docs to refer to the new unprefixed component;
1. Add the component storybook `id` (get it from the storybook URL) to the `PREVIOUSLY_EXPERIMENTAL_COMPONENTS` const array in `manager-head.html` so that old experimental story paths are redirected to the new one;
1. Add a changelog for the change.
1. Make the component public if needed, and update consumers to import it directly instead of using the private API.


- [x] Add changelog

fullofcaffeine committed:
Remove "experimental" designation for `ProgressBar`

fullofcaffeine committed:
Add optional width prop to override/set the progress bar track container

mirka commented:
The only thing I'm concerned about is the hardcoded width. If it's ok with @WordPress/gutenberg-design, I'd like to remove the width constraints so consumers can more easily set their own width using a wrapper div or something. In general, we are designing our components (especially input controls) to have an unconstrained width by default. However, I do understand that a progress bar may have different concerns, so let us know if this would be a problem.

jasmussen commented:
Can there be a default width, and a prop to override it? I think of this similar to the spinner, where it's designed to be a particular size for consistency, but where there's [a size prop](https://wordpress.github.io/gutenberg/?path=/docs/components-spinner--docs).

fullofcaffeine commented:
@mirka Are you referring to the `INDETERMINATE_TRACK_WIDTH` currently set to `50`? I don't know much about this component's API, but I can try adding the changes suggested by Joen. Let me know what you think would be the best path forward here and I'm happy to help! Thanks!

tyxla commented:
> @mirka Are you referring to the `INDETERMINATE_TRACK_WIDTH` currently set to `50`? I don't know much about this component's API, but I can try adding the changes suggested by Joen. Let me know what you think would be the best path forward here and I'm happy to help! Thanks!

If I understand correctly, @mirka referees not the progress bar track width, but rather to the entire progress bar component width (which is currently set as 160px through `max-width`):

https://github.com/WordPress/gutenberg/blob/2d33a9ed2e91efacedcec877f149061b1195c11d/packages/components/src/progress-bar/styles.ts#L28

> Can there be a default width, and a prop to override it? I think of this similar to the spinner, where it's designed to be a particular size for consistency, but where there's [a size prop](https://wordpress.github.io/gutenberg/?path=/docs/components-spinner--docs).

I'd be fine with a width prop, and then keeping the `160px` there as a reasonable default. We might want to think about how to make it fluid though and occupy the entire available width, in case that's what the 3rd party dev expects.

fullofcaffeine commented:
I folks 👋🏻 - I added an optional `width` prop, a `number` that gets interpolated as a string in the style and is assumed to use the `px` unit in the CSS property:

<video src="https://github.com/WordPress/gutenberg/assets/81248/059ec5c0-a75d-4541-93b3-620ec4f9d857" controls="controls" style="max-width: 730px;"></video>

I also added a new story to illustrate a progress bar with a custom width, but please let me know if the default story suffices.

fullofcaffeine commented on packages/components/CHANGELOG.md:
```diff
@@ -2,6 +2,8 @@
 
 ## Unreleased
 
+- `ProgressBar`: Remove "experimental" designation for `ProgressBar`, add optional custom `width` prop ([#61062](https://github.com/WordPress/gutenberg/pull/61062)).
```
Should we keep the "remove" experimental designation in the changelog or just list the `width` change? 🤔

tyxla replied:
Using a comma in the CHANGELOG is a good indication that we should have split the PR into 2 or more PRs. Have you considered moving the `width` feature (or the experimental stabilization) to a separate PR?

fullofcaffeine replied:
In this case, stabilizing/removing the experimental designation for this component is a very small change related to the storybook category, so I think we can add the API improvement as part of this PR, too. Also, by looking at other similar PRs, I can see that we shouldn't add the "remove experimental designation/stabilize" in the changelog, as it's considered an internal change. I'm happy to split, though. 

One question I have for @mirka is if we should also start exporting this component as part of this changeset to allow external users to start consuming it. I guess we should?

mirka replied:
> we should also start exporting this component as part of this changeset to allow external users to start consuming it

Oh, you're right. We need to remove the private API locking and export the component publicly.

FWIW I do agree with @tyxla that the width changes and the de-experimentalization are ideally separate PRs, if it isn't too much work.

tyxla commented on packages/components/src/progress-bar/stories/index.story.tsx:
```diff
@@ -10,9 +10,10 @@ import { ProgressBar } from '..';
+	title: 'Components/ProgressBar',
 	argTypes: {
 		value: { control: { type: 'number', min: 0, max: 100, step: 1 } },
+		width: { control: { type: 'number', min: 0, step: 10 } },
```
A `step` of `10` for `width` when we might want to test more precise values feels off.

tyxla replied:
Also, should we set a `max` still? Storybook limits the container to 1000px or something, so perhaps we should have a default max width in the story.

tyxla commented on packages/components/src/progress-bar/stories/index.story.tsx:
```diff
@@ -30,3 +31,9 @@ const Template: StoryFn< typeof ProgressBar > = ( { ...args } ) => {
 Default.args = {};
+
+// The ProgressBar width can be overriden (in pixels) and it will expand to that size if
+// the parent has enough horizontal space. It is interpolated into the `width` property
```
I guess you mean `prop`?

```suggestion
// the parent has enough horizontal space. It is interpolated into the `width` prop
```

fullofcaffeine replied:
It's the same but unabbreviated, but I get your point on leveraging React's vocab 👍🏻

tyxla commented on packages/components/src/progress-bar/index.tsx:
```diff
@@ -20,11 +20,11 @@ function UnforwardedProgressBar(
 	ref: ForwardedRef< HTMLProgressElement >
 ) {
-	const { className, value, ...progressProps } = props;
+	const { className, value, width, ...progressProps } = props;
```
I have a doubt with the prop naming. The `width` name sets the expectation that the progress bar will have a certain CSS `width`, which isn't the case. Imagine you provide `width: 400`, then if the container is 300px, the progress bar will be smaller than that. In other words, the API (`width` doesn't necessarily correspond to what happens under the hood (`max-width`). Should we rename the prop to `max-width` to make this behavior clearer?

fullofcaffeine replied:
Yeah, I thought about that. But in the end, it does set the width, just that it will fit the container. I can rename it to `max-width` if it makes it more clear/explicit, np 👍🏻

tyxla commented on packages/components/src/progress-bar/stories/index.story.tsx:
```diff
@@ -30,3 +31,9 @@ const Template: StoryFn< typeof ProgressBar > = ( { ...args } ) => {
 export const Default: StoryFn< typeof ProgressBar > = Template.bind( {} );
 Default.args = {};
+
+// The ProgressBar width can be overriden (in pixels) and it will expand to that size if
```
Is this the wrong place to document this? If the `width` prop is properly documented, it shouldn't be necessary to add a comment in this story IMHO.

fullofcaffeine replied:
Yeah, fair point.

tyxla commented on packages/components/src/progress-bar/index.tsx:
```diff
@@ -20,11 +20,11 @@ function UnforwardedProgressBar(
 	ref: ForwardedRef< HTMLProgressElement >
 ) {
-	const { className, value, ...progressProps } = props;
+	const { className, value, width, ...progressProps } = props;
```
This prop also needs to be added to the component README.

fullofcaffeine replied:
Ah true, thanks for pointing that out!

tyxla commented on packages/components/src/progress-bar/types.ts:
```diff
@@ -4,6 +4,12 @@ export type ProgressBarProps = {
 	value?: number;
 
+	/**
+	 * An optional width of the progress bar (in pixels) for the progress bar
```
The type docs may be yet another place to clarify that it's actually a max width.

tyxla commented on packages/components/src/progress-bar/styles.ts:
```diff
@@ -21,11 +21,12 @@ const animateProgressBar = keyframes( {
 	overflow: hidden;
 	width: 100%;
-	max-width: 160px;
+	max-width: ${ ( { width } ) =>
+		width ? width : CONFIG.progressBarWidth }px;
```
It may be limiting for some use cases that we're forcing the use of pixel units. If we build in the unit in the prop (and the config value), we can allow 3rd party developers to provide their unit of choice.

fullofcaffeine replied:
I followed the same pattern as the one used in the spinner component, but I agree. It might not be a good API after all.

tyxla commented on packages/components/src/progress-bar/test/index.tsx:
```diff
@@ -79,4 +80,27 @@ describe( 'ProgressBar', () => {
+	it( 'should use default width if width for the track if the `width` prop is not passed', () => {
+		const { container } = render( <ProgressBar /> );
+
+		// eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
```
Like in the pre-existing tests, it may be worth keeping the explanatory comment as to why we're accessing the `track` with inaccessible selectors.

fullofcaffeine replied:
Sure 👍🏻

tyxla commented on packages/components/src/progress-bar/test/index.tsx:
```diff
@@ -79,4 +80,27 @@ describe( 'ProgressBar', () => {
 		expect( screen.getByRole( 'progressbar' ) ).toHaveStyle( style );
 	} );
+
+	it( 'should use default width if width for the track if the `width` prop is not passed', () => {
```
```suggestion
	it( 'should use default width if the `width` prop is not passed', () => {
```

fullofcaffeine replied:
Shouldn't we be specific about what element we're setting the width for in the test? There's also the width for the inner bar.

tyxla commented on packages/components/src/progress-bar/test/index.tsx:
```diff
@@ -79,4 +80,27 @@ describe( 'ProgressBar', () => {
+		} );
+	} );
+
+	it( 'should apply custom width to tthe track if the `width` prop is passed', () => {
```
```suggestion
	it( 'should apply custom width to the track if the `width` prop is passed', () => {
```

fullofcaffeine replied:
Ops, my bad.

tyxla commented on packages/components/src/utils/config-values.js:
```diff
@@ -47,6 +47,7 @@ export default Object.assign( {}, CONTROL_PROPS, TOGGLE_GROUP_CONTROL_PROPS, {
 	borderWidthFocus: '1.5px',
 	borderWidthTab: '4px',
 	spinnerSize: 16,
+	progressBarWidth: 160,
```
Two potential confusion points with naming here:

- Above, I mentioned the one that might be confusing because of `width` vs `max-width`. This also applies here
- We don't really use this for width, but rather for `default` width. That being said, this config value will be overridden by the prop too, so it might be useful to make it clear that it's a default (max) width.

fullofcaffeine replied:
> Above, I mentioned the one that might be confusing because of width vs max-width. This also applies here
👍🏻 

> We don't really use this for width, but rather for default width. That being said, this config value will be overridden by the prop too, so it might be useful to make it clear that it's a default (max) width.

Cool, I'll rename it accordingly.

tyxla commented on packages/components/src/progress-bar/styles.ts:
```diff
@@ -21,11 +21,12 @@ const animateProgressBar = keyframes( {
 export const INDETERMINATE_TRACK_WIDTH = 50;
 
-export const Track = styled.div`
+export const Track = styled.div< { width?: number } >`
```
Does it make sense to move this to its own `TrackProps` type and use those in the `ProgressBarProps` type as well? It could potentially help make an easier distinction of which types belong where.

fullofcaffeine replied:
I can do it, but would it be worth it to create a whole new type for a single prop type? I think this simple and explicit.

tyxla reviewed (commented):
Nice work!

I believe we're addressing 2 problems here; I'd prefer removing the "experimental" designation in a separate PR. 

I also left some questions and feedback, let me know what you think.

mirka commented:
Apologies for being late to the party, but I'm not sure how I feel about a `width` prop. It would be more straightforward for consumers to control that with a wrapper div or their own CSS class.

I'm also unsure about this 160px default we're setting, because we don't have a ton of in-app usage yet and it might be too early to be prescriptive about a default. It really must depend on the context it's used in, and we don't really have much context yet.

I wonder if we can start out by suggesting a default width via documentation?

fullofcaffeine commented:
> Apologies for being late to the party, but I'm not sure how I feel about a width prop. It would be more straightforward for consumers to control that with a wrapper div or their own CSS class.

Yeah, I can see a few problems with it, the biggest one being that it's currently tied to `px` units. 

> I wonder if we can start out by suggesting a default width via documentation?

Do you mean removing the default `width` and keeping `max-width: 100%` and documenting that the consumer should wrap it in a container div and set a default width (so that the component will expand to take it and allow the user to effectively control the width this way)?

Thanks for taking the time to review this!

fullofcaffeine committed:
Revert "Add optional width prop to override/set the progress bar track container"

This reverts commit b1fe7cdf9e1739c8a17cb65dc571841eaa4bbf2c.

fullofcaffeine committed:
Keep an unconstrained width by default, while allowing consumers to override it with CSS

fullofcaffeine commented:
> Apologies for being late to the party, but I'm not sure how I feel about a width prop. It would be more straightforward for consumers to control that with a wrapper div or their own CSS class.

@mirka I simplified the implementation as you suggested - it now relies on the immediate container `width` for its width (since it has `max-width: 100%`):



<video src="https://github.com/WordPress/gutenberg/assets/81248/9ff3202f-5e90-42b3-b01f-fdb47573bce0" controls="controls" style="max-width: 730px;"></video>


I updated the docs and added a story that allows the user to change the parent element width to see how it affects the actual progress bar size. I'm not sure that's the right place for a "parent width" control, but I think it works nicely to document it and allow the user to see it in practice. Let me know if I got your right and what you think about it.

Thanks!

tyxla reviewed (commented):
Have you had a chance to test the existing `<ProgressBar />` use cases? 

There is at least one regression after removing the `max-width`, in the font library.

Before:

https://github.com/WordPress/gutenberg/assets/8436925/ae6748db-c00a-4ab5-ae1c-1c160f38e1bc



After:



https://github.com/WordPress/gutenberg/assets/8436925/76fac014-b86a-4e58-b15a-ebe8f015c5a6

mirka commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,27 @@
 
 Supports two modes: determinate and indeterminate. A progress bar is determinate when a specific progress value has been specified (from 0 to 100), and indeterminate when a value hasn't been specified.
 
+## Usage
```
Looks good! I noticed this component doesn't have a JSDoc description yet. Let's use this opportunity to add one ([example](https://github.com/WordPress/gutenberg/blob/811261494f035bd94c25cee4ae19e2159c147b72/packages/components/src/divider/component.tsx#L31-L52)).

fullofcaffeine replied:
Ah nice, thanks for pointing that out 👍🏻

fullofcaffeine replied:
Improved the docs in https://github.com/WordPress/gutenberg/pull/61062/commits/a5338e2f038c47f54463b8ee4e4742a46edf15cf and added the jsdoc as part of it.

mirka commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,27 @@
 
+## Usage
+
+Render a ProgressBar.
```
```suggestion
```

Maybe we can remove this line, because we already have a one-line description in the first section ("A simple horizontal progress bar component").

fullofcaffeine commented:
> Have you had a chance to test the existing <ProgressBar /> use cases?

Thanks for pointing that out! I'll fix that.

mirka commented on packages/components/src/progress-bar/stories/index.story.tsx:
```diff
@@ -24,9 +24,36 @@ const meta: Meta< typeof ProgressBar > = {
 export const Default: StoryFn< typeof ProgressBar > = Template.bind( {} );
 Default.args = {};
+
+export const ParentWidth: StoryFn< { parentWidth: string } > = ( {
```
> I'm not sure that's the right place for a "parent width" control (considering it's not a prop of the actual component), but I think it works nicely to show/document it and allow the user to see it in practice.

Your concerns are spot on — in our Storybook, we've decided to have our controls maintain a strict 1:1 relationship with the actual props, for clarity and consistency. So we don't want a control that's not an actual prop.

But I do think there should be a story like this to show the suggested default width (160px).

fullofcaffeine replied:
OK, if the goal is to just show a default width, I'll simplify the story to just use a custom `className` that sets it to 160px. We can then improve the docs to show the different ways of setting the width.

fullofcaffeine replied:
Done in https://github.com/WordPress/gutenberg/pull/61062/commits/5e03b5254ea2dbfcfce6e37a943fde492c2af7c5. It now looks like this:

<img width="1493" alt="Screenshot 2024-05-23 at 12 35 39 p m" src="https://github.com/WordPress/gutenberg/assets/81248/063f00e5-9874-49f7-a094-6a3e15f1cb9a">

mirka commented on packages/components/src/progress-bar/stories/index.story.tsx:
```diff
@@ -24,9 +24,36 @@ const meta: Meta< typeof ProgressBar > = {
+	parentWidth: { control: { type: 'text' }, name: 'parent width' },
+};
+
+ParentWidth.storyName = 'With Specific Width';
```
Nit: The conventions we've been using here are to just make the export const name the story name (especially when you don't need any special characters in the name). Also the story description can be defined as a JSDoc.

Example:

https://github.com/WordPress/gutenberg/blob/5fbeee2b829d98e327d9a7444ccf94f99280b964/packages/components/src/base-control/stories/index.story.tsx#L50-L57

fullofcaffeine replied:
Thanks for clarifying that 👍🏻

fullofcaffeine replied:
Done in https://github.com/WordPress/gutenberg/pull/61062/commits/5e03b5254ea2dbfcfce6e37a943fde492c2af7c5.

mirka commented on packages/components/src/progress-bar/stories/index.story.tsx:
```diff
@@ -24,9 +24,36 @@ const meta: Meta< typeof ProgressBar > = {
+	controls: { sort: 'none' },
+	docs: {
+		description: {
+			story: 'This story demonstrates how the ProgressBar takes the width of its parent container. You can edit the parent\'s width using the "Parent element width" control.',
+		},
```
Most of our components have unconstrained widths anyway, so it isn't that important to demonstrate _how_ to set a custom width here. Instead, I'm thinking we should highlight that we have a **suggested width**, mostly for Gutenberg app devs.

Maybe a description like "For most screens with a wide enough viewport, we currently suggest a max width of 160px"?

fullofcaffeine replied:
Cool, makes sense!

fullofcaffeine replied:
Done in https://github.com/WordPress/gutenberg/pull/61062/commits/5f3567e637f7208a160d48a1fd17d99289c95da3.

fullofcaffeine commented on packages/edit-site/src/components/canvas-loader/index.js:
```diff
@@ -34,7 +34,9 @@ export default function CanvasLoader( { id } ) {
 		<div className="edit-site-canvas-loader">
 			<Theme accent={ indicatorColor } background={ backgroundColor }>
-				<ProgressBar id={ id } max={ total } value={ elapsed } />
+				<div style={ { width: '160px' } }>
```
I don't like having to forcedly add these wrapper divs in the current consumers, I can't avoid thinking a self-contained default would be better here, but I'm looking forward to reading your thoughts. However, it does fix the issue [here](https://github.com/WordPress/gutenberg/pull/61062#pullrequestreview-2073018536) by making it look exactly like the current version on `trunk`.

mirka replied:
Doesn't necessarily have to be a wrapper div. Another standard way would be to pass your own class name:

```
.my-progress-bar {
  width: 160px;
}

<ProgressBar className="my-progress-bar" />
```

In either case I would prefer a CSS class rather than an inline style, since there is no reason for this to be inline.

fullofcaffeine replied:
Cool, will move it over to a CSS class there 👍🏻

fullofcaffeine replied:
Done in https://github.com/WordPress/gutenberg/pull/61062/commits/bc95336241c9240c4820d61703ac9052f652d896.

mirka reviewed (commented):
Honestly I'm more comfortable with this `width` approach at the moment, and it is more congruous with our other components. Other reviewers, feel free to chime in if you have any concerns with it. I think we can revisit the default width strategy once we have a sufficient amount of usage.

mirka commented on packages/components/src/progress-bar/stories/index.story.tsx:
```diff
@@ -24,9 +24,35 @@ const meta: Meta< typeof ProgressBar > = {
+WithDefaultSuggestedWidth.parameters = {
+	docs: {
+		description: {
+			story: 'For most screens with a wide-enough viewport, we recommend a maximum width of 160px. You can set a custom width by passing a CSS class via the `className` prop (depicted below) or by setting the width of the parent container.',
```
We can write this description as a doc comment on the story export ([like this](https://github.com/WordPress/gutenberg/pull/61062#discussion_r1612063348)), instead of in this `parameters.docs.description.story`. It's quite convenient!

fullofcaffeine replied:
Oh ok! I'll add it as a jsdoc there then 👍🏻

mirka commented on packages/components/src/progress-bar/index.tsx:
````diff
@@ -46,6 +46,53 @@ function UnforwardedProgressBar(
+ *
+ * ```jsx
+ * import { ProgressBar } from '@wordpress/components';
+ *
+ * const MyLoadingComponent = () => {
+ *   return <ProgressBar />
+ * }
+ * ```
+ *
+ * The ProgressBar will expand to take all the available horizontal space of its immediate parent container element. To control its width, you can:
+ *
+ * Pass a custom CSS `className` that takes care of setting the `width`:
+ *
+ * ```css
+ * .my-css-class {
+ *   width: 160px;
+ * }
+ * ```
+ *
+ * ```jsx
+ * import { ProgressBar } from '@wordpress/components';
+ *
+ * const MyLoadingComponent = () => {
+ *   return <ProgressBar className="my-css-class" />;
+ * };
+ * ```
+ *
+ * Wrap it in a container element (e.g `<div>`) that has a `width` specified:
+ *
+ * ```jsx
+ * import { ProgressBar } from '@wordpress/components';
+ *
+ * const MyLoadingComponent = ( props ) => {
+ *   return (
+ *     <div style={ { width: '160px' } }>
+ *       <ProgressBar />
+ *     </div>
+ *   );
+ * };
+ * ```
````
No hard rules, but in general we've been putting about this much in the component export JSDocs. A brief description, and a snippet that shows 1.) how to import it, and 2.) a basic usage example.

The main consumption points for this JSDoc data are IDE IntelliSense, and the Storybook Docs view (shown right below the default story).

```diff
 * A simple horizontal progress bar component.
 *
 * Supports two modes: determinate and indeterminate. A progress bar is determinate when a specific progress value has been specified (from 0 to 100), and indeterminate when a value hasn't been specified.
 *
-  * ## Usage
-  *
 * ```jsx
 * import { ProgressBar } from '@wordpress/components';
 *
 * const MyLoadingComponent = () => {
 *   return <ProgressBar />
 * }
 * ```
-  *
-  * The ProgressBar will expand to take all the available horizontal space of its immediate parent container element. To control its width, you can:
-  *
-  * Pass a custom CSS `className` that takes care of setting the `width`:
-  *
-  * ```css
-  * .my-css-class {
-  *   width: 160px;
-  * }
-  * ```
-  *
-  * ```jsx
-  * import { ProgressBar } from '@wordpress/components';
-  *
-  * const MyLoadingComponent = () => {
-  *   return <ProgressBar className="my-css-class" />;
-  * };
-  * ```
-  *
-  * Wrap it in a container element (e.g `<div>`) that has a `width` specified:
-  *
-  * ```jsx
-  * import { ProgressBar } from '@wordpress/components';
-  *
-  * const MyLoadingComponent = ( props ) => {
-  *   return (
-  *     <div style={ { width: '160px' } }>
-  *       <ProgressBar />
-  *     </div>
-  *   );
-  * };
-  * ```
```

fullofcaffeine replied:
Ah cool, thanks for clarifying that, that makes sense 👍🏻

tyxla commented on packages/edit-site/src/components/canvas-loader/style.scss:
```diff
@@ -16,6 +16,10 @@
 	& > div {
```
What is this style used for? Perhaps for the progress bar? IF not, couldn't we reuse it for the progress bar?

fullofcaffeine replied:
Oh, you're right, this was probably already there to set the width to the bar (though it was redundant), totally missed it 🤦🏻 

I just discussed this with Lena, and we agreed to roll back to a default behavior of having a default `max-width` + a prop that makes the width "unconstrained." Will work on it in a separate PR.

tyxla commented on packages/edit-site/src/components/global-styles/font-library-modal/style.scss:
```diff
@@ -48,6 +48,10 @@ $footer-height: 70px;
 		padding-top: $header-height + $grid-unit-15 + $grid-unit-60;
+
+		& > .font-library-modal__progressbar {
+			width: 160px;
```
What I don't like about this approach is that we have to declare it everywhere. Theoretically, if we have 20 places that use a progress bar, we need this style up to 20 times in 20 different locations. This doesn't seem sustainable and feels repetitive - can it be avoided?

fullofcaffeine replied:
Yes, I can understand the concern here. Maybe we could keep the default `width` of `160px` and allow the user to override with a CSS class? @mirka what do you think?

tyxla commented on packages/edit-site/src/components/canvas-loader/style.scss:
```diff
@@ -16,6 +16,10 @@
 	}
+
+	& > .edit-site-canvas-loader__progressbar {
+		width: 160px;
```
Is there a good reason to prefer `width` to `max-width`?

fullofcaffeine replied:
Actually no, we should indeed probably use `max-width` so that it adapts to smaller viewports. Thanks for catching that 👍🏻

tyxla commented on packages/edit-site/src/components/canvas-loader/style.scss:
```diff
@@ -16,6 +16,10 @@
 	}
+
+	& > .edit-site-canvas-loader__progressbar {
+		width: 160px;
```
I still think the work to remove the width still needs too be in a separate PR.

fullofcaffeine replied:
If the work to remove the width is considered to be separate from the stabilization work, then yes. I was under the impression it was part of it as per Lena's comment [here](https://github.com/WordPress/gutenberg/pull/61062#issuecomment-2098373670). Happy to split as well 👍🏻

tyxla commented on packages/components/src/progress-bar/stories/index.story.tsx:
```diff
@@ -24,9 +24,35 @@ const meta: Meta< typeof ProgressBar > = {
+} ) => (
+	<>
+		<style>
+			{ `
+				.progressbar-story-custom-width {
+					width: 160px;
+				}
+			` }
```
I must admit this is a very odd way to add a style. We either add it inline, or in a separate file, but this feels odd and it's [not really supported before React 19](https://react.dev/reference/react-dom/components/style) (we're still on v18.3.1).

mirka replied:
Fair. In Storybook snippets we tend to use inline for simplicity, even though it's not always the best way in production.

I don't have a strong opinion about this. We don't necessarily have to document "the correct way to set a width", since that kind of thing hasn't really been a source of confusion. I'm fine as long as we showcase the suggested width.

tyxla replied:
Less advanced folks tend to use those examples in 3rd party extension code, so if we suggest it in the docs, it increases the chance for them to think it's the best way to do things, and it will end up in their codebase.

fullofcaffeine replied:
Yeah, that's a fair point. I'll simplify that, but first, we also need to decide about [this](https://github.com/WordPress/gutenberg/pull/61062#discussion_r1613725251).

tyxla commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,51 @@
+		<div style={ { width: '160px' } }>
+			<ProgressBar />
+		</div>
+	);
```
Nit: The code spacing seems off here.

tyxla commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,51 @@
 # ProgressBar
 
-<div class="callout callout-alert">
-This feature is still experimental. “Experimental” means this is an early implementation subject to drastic and breaking changes.
-</div>
```
I think the stabilization needs its own CHANGELOG entry.

fullofcaffeine replied:
I thought it was an internal thing and shouldn't be part of the changelog.

tyxla commented on packages/components/src/progress-bar/styles.ts:
```diff
@@ -25,7 +25,6 @@ export const Track = styled.div`
 	position: relative;
 	overflow: hidden;
 	width: 100%;
-	max-width: 160px;
```
This needs its own CHANGELOG entry as well.

tyxla commented on packages/components/src/progress-bar/index.tsx:
```diff
@@ -46,6 +46,53 @@ function UnforwardedProgressBar(
 }
 
+/**
+ * A simple horizontal progress bar component.
```
I've always felt we duplicate so much README and add overhead for maintaining these docs in the future 😕 I wonder if we can improve/automate that somehow - perhaps a command to allow the README to inherit content from here?

mirka replied:
My initial plans were to abandon the manual readmes altogether once the Storybook was "complete" with all the type data and documentation. I do consider it pretty much complete now. But @DaniGuardiola mentioned some kind of TSDoc to markdown docgen he has up his sleeve, so that could be another option.

DaniGuardiola replied:
ref: https://github.com/Create-Inc/archetype/?tab=readme-ov-file#createincarchetype-docgen

tyxla replied:
I see. It would be fun to try something like this out here @DaniGuardiola 👍

fullofcaffeine replied:
Yeah, I share the non-DRY feelings here, too. A way to extract and keep everything from/in a central place would be nice 👍🏻

tyxla reviewed (commented):
I still think we're doing too many things in this single PR. It claims to stabilize the component, but it also introduces changes to the way it's consumed externally (the width). I still believe these should be separate.

fullofcaffeine committed:
Revert "Keep an unconstrained width by default, while allowing consumers to override it with CSS"

This reverts commit 64c72e2fe0ad5fb008b0c5706b6ac09314eccdab.

fullofcaffeine committed:
Make the component public

fullofcaffeine committed:
Update consumers to import the public component

fullofcaffeine committed:
Expose docs for the now public ProgressBar component

fullofcaffeine committed:
Update CHANGELOG

fullofcaffeine committed:
Merge branch 'trunk' into remove/wp-components-experimental-designation-for-progressbar

fullofcaffeine committed:
Run docs:build after docs/manifest.js change

fullofcaffeine committed:
Merge branch 'trunk' into remove/wp-components-experimental-designation-for-progressbar

* trunk:
  Update Changelog for 18.4.1
  Remove package build-types directories when cleaning (#61939)
  Bump plugin version to 18.4.1

fullofcaffeine committed:
Remove remaining private usage of ProgressBar

fullofcaffeine committed:
Small formatting fix in the changelog

fullofcaffeine committed:
Add basic docs to the README

fullofcaffeine committed:
Add jsdoc

fullofcaffeine commented:
> I'll keep working on the width behavior changes in a separate PR. 

See: https://github.com/WordPress/gutenberg/pull/61976.

tyxla commented on packages/components/CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
+### Enhancements
+
+- Components: Make the `ProgressBar` public ([#61062](https://github.com/WordPress/gutenberg/pull/61062)).
```
There's already an Enhancements section, we'll have to use it instead of adding a new one.

fullofcaffeine replied:
Oh, thanks for pointing that out.

fullofcaffeine replied:
Fixed in https://github.com/WordPress/gutenberg/pull/61062/commits/47183b7d3fecf89869d23b43a49865e463417fa9.

tyxla commented on packages/components/src/index.ts:
```diff
@@ -217,6 +217,7 @@ export {
 } from './higher-order/with-focus-return';
 export { default as withNotices } from './higher-order/with-notices';
 export { default as withSpokenMessages } from './higher-order/with-spoken-messages';
+export { ProgressBar } from './progress-bar';
```
This isn't really a high-order component, can you move the export above, where the rest of the regular components are? Another thing to keep in mind is that we usually maintain alphabetical order with exports here.

fullofcaffeine replied:
Ah, I didn't notice that this section was about HOCs. Will move.

fullofcaffeine replied:
Fixed in https://github.com/WordPress/gutenberg/pull/61062/commits/5ca947a879a9d0976fbd00d50728dc037c37ea27.

tyxla commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,42 @@
+import { ProgressBar } from '@wordpress/components';
+
+const MyLoadingComponent = () => {
+  return <ProgressBar />
```
We'll need a semicolon and use tabs for this to be valid syntax according to ESLint and Prettier:

```suggestion
	return <ProgressBar />;
```

fullofcaffeine replied:
Sounds good. As a related question, do you have ESLint/prettier activated for JSX code blocks in markdown?

tyxla replied:
Yup, I do.

tyxla commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,42 @@
+import { ProgressBar } from '@wordpress/components';
+
+const MyLoadingComponent = ( { progress } ) => {
+  return <ProgressBar value={ progress } />
```
```suggestion
	return <ProgressBar value={ progress } />;
```

tyxla commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,42 @@
+import { ProgressBar } from '@wordpress/components';
+
+const MyLoadingComponent = () => {
+  return <ProgressBar className='my-custom-css-class' />
```
In addition to the missing semicolon, note that we generally use double quotes for string literal props.

```suggestion
	return <ProgressBar className="my-custom-css-class" />;
```

tyxla commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,42 @@
+
+const MyLoadingComponent = () => {
+  return <ProgressBar />
+}
```
When declaring a constant, you also need a semicolon at the end:

```suggestion
};
```

Won't be necessary if we use the hoisted function syntax.

fullofcaffeine replied:
Yeah, my bad; I don't have syntax high-checking for jsx code blocks, so this slipped through the cracks. Thanks for spotting those.

tyxla commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,42 @@
+
+const MyLoadingComponent = ( { progress } ) => {
+  return <ProgressBar value={ progress } />
+}
```
```suggestion
};
```

tyxla commented on packages/components/src/progress-bar/README.md:
```diff
@@ -1,13 +1,42 @@
+
+const MyLoadingComponent = () => {
+  return <ProgressBar className='my-custom-css-class' />
+}
```
```suggestion
};
```

tyxla commented on packages/components/src/progress-bar/README.md:
````diff
@@ -1,13 +1,42 @@
+}
+```
+
+You can also make it controlled by passing a `value` (from 0 to 100) representing the progress:
````
```suggestion
You can also make it determinate by passing a `value` (from 0 to 100) representing the progress:
```

tyxla commented on packages/components/src/progress-bar/index.tsx:
````diff
@@ -44,6 +44,18 @@ function UnforwardedProgressBar(
+ * ```jsx
+ * import { ProgressBar } from '@wordpress/components';
+ *
+ * const MyLoadingComponent = () => {
+ *  return <ProgressBar />
+ * }
````
Let's fix this syntax as well:

```suggestion
 * const MyLoadingComponent = () => {
 * 	return <ProgressBar />;
 * };
```

tyxla reviewed (commented):
I think this is looking mostly good 👍 

A few little nits need to be ironed out, mostly docs formatting glitches.

fullofcaffeine committed:
Formatting fix in README jsx block

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
Another formatting fix in README jsx block

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
YAFF (yet another formatting fix) in the README jsx

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
Improve wording in descripton of the `value` prop in the README

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
YAFF: Missing semicolon in README jsx

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
YAFF: missing semicolon in jsx block in README

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
YAFF: missing semicolon in jsx block in README

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
YAFF: use tab instead of spaces in jsx code block in README

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
Move export out of the HOC export section

fullofcaffeine committed:
Fix CHANGELOG: Move entry to the existing enhacements section

fullofcaffeine commented:
It occurred to me that we should probably only merge this after we decide what to do with the default width/override behavior (https://github.com/WordPress/gutenberg/pull/61976). If so, both PRs should be merged in sequence. The reasoning is that if we merge this first and it takes longer to decide/merge the other, there's a (small) risk more consumers will start using it relying on the default width behavior, which might change when we merge subsequent PR.

fullofcaffeine committed:
Merge branch 'trunk' into remove/wp-components-experimental-designation-for-progressbar

jasmussen commented:
In case it helps the conversation, the progressbar was designed to always be a specific width. The fact that it can be shorter or longer than that was not actually a consideration. So a good default would be useful, even if we add flexibility on top of that.

tyxla commented on packages/components/CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@
 
 -   `Tabs`: Animate indicator ([#60560](https://github.com/WordPress/gutenberg/pull/60560)).
 -   `ComboboxControl`: Introduce Combobox expandOnFocus prop ([#61705](https://github.com/WordPress/gutenberg/pull/61705)).
+- Components: Make the `ProgressBar` public ([#61062](https://github.com/WordPress/gutenberg/pull/61062)).
```
Let's also fix the spacing: 
```suggestion
-   Components: Make the `ProgressBar` public ([#61062](https://github.com/WordPress/gutenberg/pull/61062)).
```

tyxla commented on packages/components/src/index.ts:
```diff
@@ -201,6 +201,7 @@ export {
 } from './slot-fill';
 export { default as __experimentalStyleProvider } from './style-provider';
 export { ZStack as __experimentalZStack } from './z-stack';
+export { ProgressBar } from './progress-bar';
```
As I also suggested earlier [here](https://github.com/WordPress/gutenberg/pull/61062#discussion_r1615858936), can we move this up to follow the pre-existing alphabetical order?

fullofcaffeine replied:
Ah, sure 👍🏻 will keep that in mind as well. Thanks for the reminder.

fullofcaffeine replied:
Done: https://github.com/WordPress/gutenberg/pull/61062/commits/43e1a57eb2bc2cea2f7bb6639f293de4e700c1f0.

tyxla reviewed (commented):
A couple of final nits, other than that, looking good and can be merged once the `width` debate is finalized.

fullofcaffeine committed:
Spacing fix

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
Sort import alphabetically

tyxla commented on packages/components/src/index.ts:
```diff
@@ -132,6 +132,7 @@ export { default as PanelHeader } from './panel/header';
 export { default as PanelRow } from './panel/row';
 export { default as Placeholder } from './placeholder';
 export { default as Popover } from './popover';
+export { ProgressBar } from './progress-bar';
```
Maybe we could use `default as ProgressBar` to be consistent with the rest.

fullofcaffeine replied:
Sure, it'd be easier on the eyes.

fullofcaffeine replied:
Done in https://github.com/WordPress/gutenberg/pull/61062/commits/2f978f4af873993a7aa2bea0c8494bb368d2453a.

tyxla commented on packages/components/CHANGELOG.md:
```diff
@@ -11,6 +11,7 @@
 
 -   `Tabs`: Animate indicator ([#60560](https://github.com/WordPress/gutenberg/pull/60560)).
 -   `ComboboxControl`: Introduce Combobox expandOnFocus prop ([#61705](https://github.com/WordPress/gutenberg/pull/61705)).
+-   Components: Make the `ProgressBar` public ([#61062](https://github.com/WordPress/gutenberg/pull/61062)).
```
`Components` is a redundant prefix, since this entire package is about components. Here's a suggestion to remove the redundancy:

```suggestion
-   `ProgressBar`: Expose as public API ([#61062](https://github.com/WordPress/gutenberg/pull/61062)).
```

fullofcaffeine replied:
Sounds good 👍🏻

tyxla reviewed (approved):
Ready to go when stable 🚀 

Left a few more optional nits.

fullofcaffeine committed:
Simplify CHANGELOG entry

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

fullofcaffeine committed:
Re-export as default for visual consistency

fullofcaffeine committed:
Merge branch 'trunk' into remove/wp-components-experimental-designation-for-progressbar

fullofcaffeine committed:
Merge branch 'trunk' into remove/wp-components-experimental-designation-for-progressbar

fullofcaffeine commented:
There's a Playwright test [failing](https://github.com/WordPress/gutenberg/actions/runs/9290331396/job/25567476240?pr=61062), but it's [also failing](https://github.com/WordPress/gutenberg/actions/runs/9289883244/job/25564822251) on `trunk`, so not related to this changeset.

fullofcaffeine committed:
Improve jsdoc comment: make it fit within ~80chars of char lenght and fix comment opening

fullofcaffeine commented:
I forgot to paste the contributor text to the commit description; sorry! The PR had a lengthy list of discussions/messages and missed the [warning](https://github.com/WordPress/gutenberg/pull/61062#issuecomment-2123645210) at the top. Will a private not to forget for the following PRs.
